File: backend/src/services/qa.js
import axios from "axios";
import { logger } from "../config/logger.js";
import { Resource } from "../models/Resource.js";

const AI_SERVICE_URL = process.env.AI_SERVICE_URL || "http://localhost:8000";

/**
 * Service for QA operations (question answering with RAG)
 */

export const qaService = {
  /**
   * Ask a question and get an answer with sources
   */
  async askQuestion(question, resourceIds = null) {
    try {
      if (!question || question.trim().length === 0) {
        throw new Error("Question cannot be empty");
      }

      if (question.length > 500) {
        throw new Error("Question too long (max 500 characters)");
      }

      logger.info(`Processing QA: ${question.substring(0, 100)}`);

      const payload = {
        question: question.trim(),
      };

      if (resourceIds && Array.isArray(resourceIds) && resourceIds.length > 0) {
        payload.resource_ids = resourceIds;
      }

      const response = await axios.post(
        `${AI_SERVICE_URL}/api/v1/qa/answer`,
        payload,
        {
          timeout: 60000, // 60s timeout for QA
        },
      );

      return {
        answer: response.data.answer,
        sources: response.data.sources || [],
        confidence: response.data.confidence || 0,
        tokens_used: response.data.tokens_used || 0,
        processing_time: response.data.processing_time_ms || 0,
        source_count:
          response.data.source_count ||
          (response.data.sources || []).length ||
          0,
        answer_mode: response.data.answer_mode || "ai",
        answer_label: response.data.answer_label || "AI Answer",
      };
    } catch (error) {
      logger.error("Error in askQuestion:", error.message);
      if (error.response?.data?.detail) {
        throw new Error(error.response.data.detail);
      }
      throw new Error(`Failed to generate answer: ${error.message}`);
    }
  },

  /**
   * Store QA interaction for analytics
   */
  async storeInteraction(userId, question, answer, sources, resourceIds = []) {
    try {
      const payload = {
        userId,
        question,
        answer,
        sources,
        resourceIds,
        processingTime: Date.now(),
      };

      await axios.post(
        `${AI_SERVICE_URL}/api/v1/qa/store-interaction`,
        payload,
        {
          timeout: 10000,
        },
      );

      logger.info(`QA interaction stored for user: ${userId}`);
    } catch (error) {
      // Non-critical error - log but don't throw
      logger.warn("Failed to store QA interaction:", error.message);
    }
  },

  /**
   * Rate a QA answer
   */
  async rateAnswer(questionId, userId, rating) {
    try {
      if (!["helpful", "not-helpful"].includes(rating)) {
        throw new Error("Rating must be 'helpful' or 'not-helpful'");
      }

      await axios.post(`${AI_SERVICE_URL}/api/v1/qa/rate`, {
        question_id: questionId,
        user_id: userId,
        rating,
      });

      logger.info(`Answer rated: ${rating}`);
    } catch (error) {
      logger.warn("Failed to rate answer:", error.message);
    }
  },

  /**
   * Get recent questions from a user
   */
  async getRecentQuestions(userId, limit = 10) {
    try {
      const response = await axios.get(
        `${AI_SERVICE_URL}/api/v1/qa/user/${userId}/history`,
        {
          params: { limit },
          timeout: 10000,
        },
      );

      return response.data.questions || [];
    } catch (error) {
      logger.warn("Failed to fetch question history:", error.message);
      return [];
    }
  },
};

File: ai-service/app/routes/qa.py
from fastapi import APIRouter, HTTPException, UploadFile, File
from pydantic import BaseModel
from typing import List, Optional
import logging
from app.services.qa import RAGService
import os
from datetime import datetime

router = APIRouter(prefix="/qa", tags=["QA"])
logger = logging.getLogger(__name__)

# Initialize RAG service
rag_service = None


# Pydantic models
class QuestionRequest(BaseModel):
    question: str
    resource_ids: Optional[List[str]] = None


class QAInteractionRequest(BaseModel):
    userId: str
    question: str
    answer: str
    sources: List[dict]
    processingTime: int
    resourceIds: List[str] = []


class RatingRequest(BaseModel):
    question_id: str
    user_id: str
    rating: str  # "helpful" or "not-helpful"


# Initialize RAG service on startup
async def initialize_rag_service():
    global rag_service
    if rag_service is None:
        from app.config.database import db as database
        rag_service = RAGService(database)
        logger.info("RAG Service initialized")


# POST /ai/qa/answer - Main QA endpoint
@router.post("/answer")
async def answer_question(
    request: QuestionRequest,
):
    """
    Process a question and return an AI-generated answer with sources.
    Uses RAG (Retrieval-Augmented Generation) pipeline.
    """
    try:
        if not request.question or len(request.question.strip()) == 0:
            raise HTTPException(status_code=400, detail="Question cannot be empty")

        if len(request.question) > 500:
            raise HTTPException(status_code=400, detail="Question too long (max 500 characters)")

        logger.info(f"Processing question: {request.question[:100]}")

        # Get or initialize RAG service
        global rag_service
        if rag_service is None:
            await initialize_rag_service()

        # Generate answer using RAG pipeline
        result = await rag_service.generate_answer(
            question=request.question,
            resource_ids=request.resource_ids,
        )

        return {
            "answer": result["answer"],
            "sources": result["sources"],
            "confidence": result.get("confidence", 0),
            "tokens_used": result.get("tokens_used", 0),
            "processing_time_ms": result.get("processing_time", 0),
            "answer_mode": result.get("answer_mode", "ai"),
            "answer_label": result.get("answer_label", "AI Answer"),
            "source_count": result.get("source_count", len(result.get("sources", []))),
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in answer_question: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail="Error generating answer. Please try again.",
        )


# POST /ai/qa/store-interaction - Store QA interaction
@router.post("/store-interaction")
async def store_interaction(
    request: QAInteractionRequest,
):
    """Store a QA interaction in the database for tracking and insights."""
    try:
        global rag_service
        if rag_service is None:
            await initialize_rag_service()

        await rag_service.store_interaction(
            user_id=request.userId,
            question=request.question,
            answer=request.answer,
            sources=request.sources,
            processing_time=request.processingTime,
            resource_ids=request.resourceIds,
        )

        return {"success": True, "message": "Interaction stored"}

    except Exception as e:
        logger.error(f"Error storing interaction: {str(e)}")
        # Don't fail - this is non-critical
        return {"success": False, "message": "Failed to store interaction"}


# GET /ai/qa/user-history - Get user's QA history
@router.get("/user-history")
async def get_user_history(
    user_id: str,
    limit: int = 10,
):
    """Retrieve a user's question history."""
    try:
        if limit > 50:
            limit = 50
        if limit < 1:
            limit = 10

        global rag_service
        if rag_service is None:
            await initialize_rag_service()

        interactions = await rag_service.get_user_history(user_id, limit)

        return {
            "user_id": user_id,
            "interactions": interactions,
        }

    except Exception as e:
        logger.error(f"Error retrieving history: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to retrieve history")


# POST /ai/qa/rate-answer - Rate an answer
@router.post("/rate-answer")
async def rate_answer(request: RatingRequest):
    """Store user rating for an answer."""
    try:
        if request.rating not in ["helpful", "not-helpful"]:
            raise HTTPException(
                status_code=400,
                detail='Rating must be "helpful" or "not-helpful"',
            )

        global rag_service
        if rag_service is None:
            await initialize_rag_service()

        await rag_service.rate_answer(
            question_id=request.question_id,
            user_id=request.user_id,
            rating=request.rating,
        )

        return {"success": True, "message": "Rating recorded"}

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error rating answer: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to save rating")


# POST /ai/qa/clear-cache - Clear cached answers (admin)
@router.post("/clear-cache")
async def clear_cache(admin_key: str = None):
    """Clear all cached QA responses. Requires admin key."""
    try:
        if admin_key != os.getenv("ADMIN_KEY"):
            raise HTTPException(status_code=401, detail="Unauthorized")

        global rag_service
        if rag_service is None:
            from app.config.database import db as database
            await initialize_rag_service(database)

        await rag_service.clear_cache()

        return {"success": True, "message": "Cache cleared"}

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error clearing cache: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to clear cache")
